DOC: Added a PR "Reviewer guidelines" document. (#17252)
numpy/numpy: doc/source/dev/reviewer_guidelines.rst

  welcoming, friendly and patient. Be `kind
  <https://youtu.be/tzFWz5fiVKU?t=49m30s>`_ to contributors.
- Do not let perfect be the enemy of the good, particularly for documentation.
  If you find yourself making many small suggestions, or being too nitpicky on
  style or grammar, consider merging the current PR when all important concerns
  are addressed. Then, either push a commit directly (if you are a maintainer)
  or open a follow-up PR yourself.
- If you need help writing replies in reviews, check out some `Standard replies
  for reviewing
  <https://scikit-learn.org/stable/developers/tips.html#saved-replies>`_.

Reviewer Checklist
==================

- Is the intended behavior clear under all conditions? Some things to watch:
   - What happens with unexpected inputs like empty arrays or nan/inf values?
   - Are axis or shape arguments tested to be `int` or `tuples`?
   - Are unusual `dtypes` tested if a function supports those?
- Should variable names be improved for clarity or consistency?
- Should comments be added, or rather removed as unhelpful or extraneous?
- Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are

>   the docstrings properly formatted?

Reviewer: On the same note, can we add a reference to the code style guidelines here? https://numpy.org/doc/stable/dev/index.html#stylistic-guidelines (but with an internal sphinx link)

Reviewer 2: done, resolving.

  project it will be <https://numpy.org/code-of-conduct/>`_: open, empathetic,
  welcoming, friendly and patient. Be `kind
  <https://youtu.be/tzFWz5fiVKU?t=49m30s>`_ to contributors.
- Do not let perfect be the enemy of the good, particularly for documentation.
  If you find yourself making many small suggestions, or being too nitpicky on
  style or grammar, consider merging the current PR when all important concerns
  are addressed. Then, either push a commit directly (if you are a maintainer)
  or open a follow-up PR yourself.
- If you need help writing replies in reviews, check out some `Standard replies
  for reviewing
  <https://scikit-learn.org/stable/developers/tips.html#saved-replies>`_.

Reviewer Checklist
==================

- Is the intended behavior clear under all conditions? Some things to watch:
   - What happens with unexpected inputs like empty arrays or nan/inf values?
   - Are axis or shape arguments tested to be `int` or `tuples`?
   - Are unusual `dtypes` tested if a function supports those?
- Should variable names be improved for clarity or consistency?
- Should comments be added, or rather removed as unhelpful or extraneous?

> - Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are

Reviewer 2: still an error in the reference here, needs a space

Reviewer 2: or maybe this does work, since you're doing it in multiple place. let me go look at the html

  welcoming, friendly and patient. Be `kind
  <https://youtu.be/tzFWz5fiVKU?t=49m30s>`_ to contributors.
- Do not let perfect be the enemy of the good, particularly for documentation.
  If you find yourself making many small suggestions, either open a PR against
  the original branch, push changes to the contributor branch, or merge the PR
  and then open a new PR against upstream.
- If you need help writing replies in reviews, check out some `Standard replies
  for reviewing
  <https://scikit-learn.org/stable/developers/tips.html#saved-replies>`_.

Reviewer Checklist
==================

- Is the intended behavior clear under all conditions (e.g. unexpected inputs
  like empty arrays or nan/inf values)?
- Is the code easy to read and low on redundancy? Should variable names be
  improved for clarity or consistency? Should comments be added? Should comments
  be removed as unhelpful or extraneous?
- Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are
  the docstrings properly formatted?
- If it is not obvious from the PR description, add a short explanation of what
  a branch did to the merge message and, if closing an issue, also add "Closes

>   #123" where 123 is the issue number.

Reviewer 2: minor: `gh-123` would be preferred

- For code changes, at least one core team member (those with commit rights)
  should review and approve all pull requests. If you are the first to review a
  PR and approve of the changes use the GitHub `approve review
  <https://help.github.com/articles/reviewing-changes-in-pull-requests/>`_ tool
  to mark it as such. If you are a subsequent reviewer, please approve the
  review or, if you have commit rights and think no more review is needed, merge

>   the PR.

Reviewer 2: This seems like a leftover from scikit-image, where they require two reviewers.

Author: Actually the way I wanted to write this is that two reviewers are not required, but if someone wants to make suggestions on an already approved PR they can (even with no commit rights). If that's not clear I can rephrase

Reviewer 2: Let me try to propose a rephrase. Like this it seems to encourage a habit I see too often, which is that all important review is done, and then someone comes by and adds one or a couple of nitpicky comments. Which then stalls the merge and is super annoying for the contributor, for no real gain. I want to make clear that's very much not desired.

Reviewer 2: Proposed change to last bullet:
```
- For code changes, at least one maintainer (i.e. someone with commit rights)
  should review and approve a pull request. If you are the first to review a
  PR and approve of the changes use the GitHub `approve review
  <https://help.github.com/articles/reviewing-changes-in-pull-requests/>`_ tool
  to mark it as such. If a PR is straightforward, for example it's a clearly
  correct bug fix, it can be merged straight away. If it's more complex or
  changes public API, please leave it open for at least a couple of days so
  other maintainers get a chance to review.
- If you are a subsequent reviewer on an already approved PR, please use the
  same review method as for a new PR (focus on the larger issues, resist the
  temptation to add only a few nitpicks).  If you have commit rights and think
  no more review is needed, merge the PR.
```

> Core team members

Reviewer 2: Can we change this to "For maintainers"? I really dislike the term "core team" (implies that some type of activities are more important than others) and want to get rid of it soon.

-----------------
  
- Make sure all automated CI tests pass before merging a PR, and that the

>   :ref:`documentation builds<building-docs>` without any errors.

Reviewer 2: missing a space between `builds` and `<`

- In case of merge conflicts, ask the PR submitter to :ref:`rebase on master
  <rebasing-on-master>`.
- For PRs that add new features or are in some way complex, wait at least a day
  or two before merging it. That way, others get a chance to comment before the
  code goes in. Consider adding it to the release notes.
- When merging contributions, a committer is responsible for ensuring that those
  meet the requirements outlined in the :ref:`Development process guidelines
  <guidelines>` for NumPy. Also, check that new features and backwards
  compatibility breaks were discussed on the `numpy-discussion mailing list
  <https://mail.python.org/mailman/listinfo/numpy-discussion>`_.
- Backports and trivial additions to finish a PR (really trivial, like a typo or

>   PEP8 fix) can be pushed directly.

Reviewer 2: Let's get rid of this bullet, we don't do that anymore.

- Squashing commits or cleaning up commit messages of a PR that you consider too
  messy is OK. Remember to retain the original author's name when doing this.
  Make sure commit messages follow the :ref:`rules for NumPy
  <writing-the-commit-message>`.
- When you want to reject a PR: if it's very obvious, you can just close it and
  explain why. If it's not, then it's a good idea to first explain why you
  think the PR is not suitable for inclusion in NumPy and then let a second
  committer comment or close.

GitHub Workflow
---------------

When reviewing pull requests, please use workflow tracking features on GitHub as
appropriate:

- After you have finished reviewing, if you want to ask for the submitter to
  make changes, change your review status to "Changes requested." This can be
  done on GitHub, PR page, Files changed tab, Review changes (button on the top
  right).
- If you're happy about the current status, mark the pull request as Approved

>   (same way as Changes requested). Alternatively (for core team members): merge

Reviewer 2: another "core team" instance here

  the pull request, if you think it is ready to be merged.

It may be helpful to have a copy of the pull request code checked out on your
own machine so that you can play with it locally. One way to do this is to
navigate to the NumPy root directory in the terminal and enter::
  

>   git fetch upstream pull/PULL_REQUEST_ID/head:NEW_BRANCH_NAME

Reviewer 2: Let's replace this bit with `gh` or `hub` CLI usage (or get rid of it).

Reviewer 2: Oh wait, that's below. I think that's so far superior that we should just delete this old hacky method.

Reviewer 2: this apparently does work, learned something. resolving.

Reviewing open pull requests (PRs) is very welcome and a valuable way to help
increase the speed at which the project moves forward. We encourage everyone to

> get involved in the review process; it's also a great way to get familiar with

Reviewer 3: ```suggestion
Reviewing open pull requests (PRs) is very welcome and helps
the project move forward. We encourage everyone to
get involved; it's also a great way to get familiar with
```

the codebase.

Who can be a reviewer?
======================

> You don't have to be a project member to contribute reviews. Consider being a
> reviewer if 
> 
> - You have contributed to the NumPy project before, and feel like you can
>   comment on technical or organizational aspects of the PR; 
> - You haven't contributed before, but are an experienced NumPy user, and see a
>   pull request that touches a funcionality that you use often;

Reviewer 3: ```suggestion
You don't have to be a project member to contribute reviews. Consider reviewing
if you're an experienced user and the request touches functionality you use often.
```
```

- You can contribute comments about content, style, organization or good

>   documentation practices.  

Reviewer 3: Not needed; review topics are discussed in detail below.

If we do not know you yet, consider introducing yourself before you start

> reviewing pull requests.

Reviewer 3: Be clearer about what "introducing yourself" means.

Communication Guidelines
========================

- Every PR, good or bad, is an act of generosity. Opening with a positive
  comment will help the author feel rewarded, and your subsequent remarks may be
  heard more clearly. You may feel good also.
- Begin if possible with the large issues, so the author knows they've been
  understood. Resist the temptation to immediately go line by line, or to open
  with small pervasive issues.
- You are the face of the project, and NumPy some time ago decided `the kind of
  project it will be <https://numpy.org/code-of-conduct/>`_: open, empathetic,
  welcoming, friendly and patient. Be `kind
  <https://youtu.be/tzFWz5fiVKU?t=49m30s>`_ to contributors.
- Do not let perfect be the enemy of the good, particularly for documentation.
  If you find yourself making many small suggestions, either open a PR against
  the original branch, push changes to the contributor branch, or merge the PR

>   and then open a new PR against upstream.

Reviewer 3: * It's not clear how these steps keep perfect from being the enemy of the good or how they move things along.
* Some of these steps are available only to maintainers

- If you need help writing replies in reviews, check out some `Standard replies
  for reviewing

>   <https://scikit-learn.org/stable/developers/tips.html#saved-replies>`_.

Reviewer 3: Let's have our own canned set. It's more respectable, it seems easy to do, and it will avoid people copying scikit-learn stuff by accident.

Reviewer 2: +1, but can be done in a follow up PR

Reviewer Checklist
==================

- Is the intended behavior clear under all conditions (e.g. unexpected inputs

>   like empty arrays or nan/inf values)?

Reviewer 3: This is too skimpy. That's ironic for an item that's urging thoroughness. We should give a fuller checklist of cases likely to be overlooked and suggestions on how to think of them generally. The list might be elsewhere and linked to here. 

We should also mention (possibly in a separate bullet) the `hypothesis` tool, for anyone who knows (or may want to learn) how to use it.

Reviewer 2: Empty arrays and inf/nan are the main special cases. Nothing else immediately comes to mind - do you have any suggestions?

Reviewer 4: Here are some more thoughts, some may be impractical/not appropriate
- axis or shape arguments should be int and tuples
- scalars and 0-d arrays
- there are techniques to use all the dtypes, which means that new ones (will we ever do another datetime?) will be added to the tests automatically when they are added to NumPy
- tests should hit the error cases in the code

Reviewer 5: object arrays are huge special cases (e.g. array containing fractions/decimals).

Reviewer 2: > object arrays are huge special cases (e.g. array containing fractions/decimals).

Yes, but do we really want to require tests with things like object arrays for decimals for all functionality? That seems like an unhealthy idea.

Reviewer 2: +1 to all dtypes and axis int/tuple. the rest seems like overdoing it for a general review recommendation

Reviewer 5: @rgommers I thought this was a "checklist" for what reviewers may want to look out for, not a strict requirement?

Reviewer 2: Yes, but inevitably newer reviewers are going to ask for that stuff if it's on this list, and for something obscure like object array of decimals that's almost never desired. So making the list overly long isn't ideal imho.

Author: Can one of you explain the technique for testing all dtypes? I'm not familiar with it and don't really know how to explain. I'd say something like:

> Instead of listing specific dtypes when testing, *use technique*. This means that when new dtypes are added to NumPy, the tests will automatically include them.

Reviewer 2: There's no standard way of grabbing a set of dtypes. Tests normally use `pytest.mark.parametrize`. For example:
```
int_types = [np.byte, np.short, np.intc, np.int_, np.longlong]
uint_types = [np.ubyte, np.ushort, np.uintc, np.uint, np.ulonglong]
float_types = [np.half, np.single, np.double, np.longdouble]
cfloat_types = [np.csingle, np.cdouble, np.clongdouble]

class TestArrayFromScalar:

    @pytest.mark.parametrize('t1', int_types + uint_types)
    @pytest.mark.parametrize('t2', int_types + uint_types + [None])
    def test_integers(self, t1, t2):
        return self._do_test(t1, t2)

    @pytest.mark.parametrize('t1', float_types)
    @pytest.mark.parametrize('t2', float_types + [None])
    def test_reals(self, t1, t2):
        return self._do_test(t1, t2)
```

The dtypes is something that requires some thought. Sometimes it uses `float32, float64`, or `"f4", "f8", ">f8",`. A search for `mark.parametrize` will show you a lot of ways of doing this.

Reviewer 2: Actually, this is not the place to explain even `parametrize`. Maybe just add: _"are unusual dtypes tested if a function supports those?"_

> - Is the code easy to read and low on redundancy? Should variable names be

Reviewer 3: We don't have to tell reviewers to complain if they can't make sense of the code. But we want them to make constructive suggestions, not just complain, and the list of failures and improvements is much larger than these two items.  Rather than opening the door for vague readability complaints, we should either cover readability comprehensively (possibly with a link), or omit talk of readability and break out the two items (variables and comments) individually

- Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are
  the docstrings properly formatted?
- If it is not obvious from the PR description, add a short explanation of what

>   a branch did to the merge message and, if closing a bug, also add "Closes

Reviewer 3: The advice on merge messages is pertinent only to maintainers.

  #123" where 123 is the issue number.

> - For code changes, at least one core team member (those with commit rights)
>   should review and approve all pull requests. If you are the first to review a

Reviewer 3: Not sure of the value of the first sentence. The only way the PR can be committed to master is if someone with commit rights does it.
```suggestion
-  If you are the first to review a
```
